packages/browser-pool/tab-as-a-container/background.js

```diff
@@ -0,0 +1,383 @@
+    details => {
+        for (const header of details.responseHeaders) {
+            if (header.name.toLowerCase() === 'set-cookie') {
+                const parts = header.value.split('\n');
+
+                // `details.tabId` === -1 when Chrome is making internal requests, such downloading a service worker.
+
+                const openerId = getOpenerId(details.tabId);
+
+                header.value = parts.map(part => {
+                    const equalsIndex = part.indexOf('=');
+                    if (equalsIndex === -1) {
+                        return `${keyFromTabId(openerId)}=${part.trimStart()}`;
+                    } else {
+                        return keyFromTabId(openerId) + part.trimStart();
+                    }
+                }).join('\n');
+            }
+        }
+
+        return {
+            responseHeaders: details.responseHeaders,
+        };
+    },
+    {urls: ['<all_urls>']},
+    webRequestPermissions.blockingResponse,
+);
+
+chrome.tabs.onRemoved.addListener(async tabId => {
+    const opener = getOpenerId(tabId);
+    translator.delete(tabId);
+
+    if (counter.has(opener)) {
+        counter.set(opener, counter.get(opener) - 1);
+
+        if (counter.get(opener) < 1) {
+            counter.delete(opener);
+        } else {
+            return;
+        }
+    }
+
+    const id = keyFromTabId(opener);
+
+    chrome.cookies.getAll({}, async cookies => {
+        await Promise.allSettled(cookies.filter(cookie => cookie.name.startsWith(id)).map(cookie => {
+            return chrome.cookies.remove({
+                name: cookie.name,
+                url: getCookieURL(cookie),
+                storeId: cookie.storeId,
+            });
+        }));
+    });
+});
+
+// Proxy per tab
+const getProxyConfiguration = (scheme, host, port) => {
+    return {
+        mode: 'fixed_servers',
+        rules: {
+            proxyForHttp: {
+                scheme,
+                host,
+                port,
+            },
+            proxyForHttps: {
+                scheme,
+                host,
+                port,
+            },
+        }
+    };
+};
+
+const localhostIpCache = new Map();
+const localHostIp = [127, 0, 0, 0];
+const getNextLocalhostIp = openerId => {
+    if (localhostIpCache.has(openerId)) {
+        return localhostIpCache.get(openerId);
+    }
+
+    const result = localHostIp.join('.');
+
+    localhostIpCache.set(openerId, result);
+
+    // 255 * 255 * 255 = 16 581 375
+    if (localHostIp[3] === 255) {
+        if (localHostIp[2] === 255) {
+            if (localHostIp[1] === 255) {
+                localHostIp[0] = 127;
+                localHostIp[1] = 0;
+                localHostIp[2] = 0;
+                localHostIp[3] = 0;
+            } else {
+                localHostIp[1]++;
+            }
+        } else {
+            localHostIp[2]++;
+        }
+    } else {
+        localHostIp[3]++;
+    }
+
+    while (localhostIpCache.length >= (1 * 255 * 255 * 255)) {
```

Comment:
```suggestion
    while (localhostIpCache.length >= (255 * 255 * 255)) {
```

Reply:
That 1 is for visibility: `127.0.0.0` with a mask of `0.255.255.255`

Reply:
I was thinking it was quite obvious, since we rotate from `127.0.0.0` to `127.255.255.255` (given the loop above). Which is... wait it's actually 256 :joy:

Reply:
Actually wait. localhost is /24 so `127.255.255.255` is broadcast. And `127.0.0.0` is interface. So I can't actually use that... Fixed! And added a comment :)